### PR TITLE
[feat] Add support for the OAR scheduler

### LINKED ISSUE #2124: Minimal support for OAR batch scheduler
Hello,

It would be nice for our work to have scheduler support for [OAR](https://oar.imag.fr/start) batch scheduler mainly used on experimentation platforms. [Grid5000](https://www.grid5000.fr/w/Hardware) network of clusters in France uses OAR as workload manager and it offers a plethora of architectures to test the codes. ReFrame would be a very appropriate solution to perform regression tests on these clusters.

I have started working on implementing the scheduler. It has very similar properties to PBS and SGE, so, I am using them as a base.

### PR DESCRIPTION
**Pull request by mahendrapaipuri**
This PR adds support to a minimal [OAR](https://oar.imag.fr/documentation) backend scheduler. 
Fixes #2124.

### CODE REVIEW

**Comment by teojgo** [comment]
Ok to test

**Comment by teojgo** [comment]
@jenkins-cscs retry all

File: reframe/core/schedulers/oar.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        options += job.sched_access + job.options + job.cli_options
+
+        for opt in options:
+            if opt.startswith('#'):
+                preamble.append(opt)
+            else:
+                preamble.append(self._format_option(opt))
+
+        # OAR starts the job in the home directory by default
+        preamble.append('cd %s' % job.workdir)
```
Comment (vkarak) [inline reframe/core/schedulers/oar.py:105]: Use f-strings here.

File: reframe/core/schedulers/oar.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        for job in jobs:
+            completed = osext.run_command(
+                f'oarstat -fj {job.jobid}'
+            )
+
+            if completed.returncode != 0:
+                raise JobSchedulerError(
+                    f'oarstat failed with exit code {completed.returncode} '
+                    f'(standard error follows):\n{completed.stderr}'
+                )
```
Comment (vkarak) [inline reframe/core/schedulers/oar.py:147]: Just use `_run_strict()` here and do not throw an extra error.

**Review by vkarak — CHANGES_REQUESTED** [review]
Thanks @mahendrapaipuri for your PR I cannot test the backend, so my comments are related to the code itself.

**Comment by vkarak** [comment]
We would also need an entry in the documentation.

File: unittests/test_schedulers.py [hunk 3]
```diff
@@ -208,6 +209,22 @@ def _expected_torque_directives(job):
...
 
+def _expected_oar_directives(job):
+    num_nodes = job.num_tasks // job.num_tasks_per_node
+    num_tasks_per_node = job.num_tasks_per_node
+    return set([
+        '#OAR -n "testjob"',
+        '#OAR -O %s' % job.stdout,
+        '#OAR -E %s' % job.stderr,
+        '#OAR -l /host=%s/core=%s,walltime=0:5:0' % (num_nodes,
+                                                     num_tasks_per_node),
```
[the remark this answers is not in the document]
Comment (mahendrapaipuri) [reply unittests/test_schedulers.py]: Fixed

File: reframe/core/schedulers/oar.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+    def emit_preamble(self, job):
+        # Same reason as oarsub, we give full path to output and error files to
+        # avoid writing them in the working dir
+        preamble = [
+            self._format_option('-n "%s"' % job.name),
+            self._format_option('-O %s' %
+                                os.path.join(job.workdir, job.stdout)),
+            self._format_option('-E %s' %
+                                os.path.join(job.workdir, job.stderr)),
```
[the remark this answers is not in the document]
Comment (mahendrapaipuri) [reply reframe/core/schedulers/oar.py]: Fixed

File: reframe/core/schedulers/oar.py (hunk 1 above)
Comment (mahendrapaipuri) [reply reframe/core/schedulers/oar.py]: Fixed

File: reframe/core/schedulers/oar.py (hunk 2 above)
Comment (mahendrapaipuri) [reply reframe/core/schedulers/oar.py]: Fixed

**Comment by mahendrapaipuri** [comment]
> We would also need an entry in the documentation.

Added an entry in the documentation. I added a line about CI image as well. I am sorry that I forgot to push that change in that previous PR.

File: docs/tutorial_tips_tricks.rst [hunk 5]
```diff
@@ -552,8 +552,10 @@ The following is an example of ``.gitlab-ci.yml`` file that does exactly that:
 It defines two stages.
 The first one, called ``generate``, will call ReFrame to generate the pipeline specification for the desired tests.
 All the usual `test selection options <manpage.html#test-filtering>`__ can be used to select specific tests.
-ReFrame will process them as usual, but instead of running the selected tests, it will generate the correct steps for running each test individually as a Gitlab job.
-We then pass the generated CI pipeline file to second phase as an artifact and we are done!
+ReFrame will process them as usual, but instead of running the selected tests, it will generate the correct steps
+for running each test individually as a Gitlab job. We then pass the generated CI pipeline file to second phase as
+an artifact and we are done! If ``image`` keyword is defined in ``.gitlab-ci.yml``, the emitted pipeline will use
+the same image as the one defined in the parent pipeline.
```
Comment (vkarak) [inline docs/tutorial_tips_tricks.rst:558]: We have a policy in the documentation to start new sentences in new lines for making diffs easier:

```suggestion
an artifact and we are done! 
If the ``image`` keyword is defined in ``.gitlab-ci.yml``, the emitted pipeline will use
the same image as the one defined in the parent pipeline.
The verbosity level of the initial invocation of ReFrame will also be retained in the invocations from the child pipeline.
```

Comment (vkarak) [reply docs/tutorial_tips_tricks.rst:558]: It'd be better if we had this doc enhancement in a separate PR.

**Review by vkarak — CHANGES_REQUESTED** [review]
These changes broke the unit tests for `test_prepare[oar]`:

```
E       assert {'#DW jobdw capacity=100GB',\n '#DW stage_in source=/foo',\n '#OAR --gres=gpu:4',\n '#OAR -E '\n '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.err',\n '#OAR -O '\n '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.out',\n '#OAR -l /host=8/core=2,walltime=0:5:0#OAR --account=spam',\n '#OAR -n "testjob"'} == {'#DW jobdw capacity=100GB',\n '#DW stage_in source=/foo',\n '#OAR --account=spam',\n '#OAR --gres=gpu:4',\n '#OAR -E '\n '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.err',\n '#OAR -O '\n '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.out',\n '#OAR -l /host=8/core=2,walltime=0:5:0',\n '#OAR -n "testjob"'}
E         Extra items in the left set:
E         '#OAR -l /host=8/core=2,walltime=0:5:0#OAR --account=spam'
E         Extra items in the right set:
E         '#OAR -l /host=8/core=2,walltime=0:5:0'
E         '#OAR --account=spam'
E         Full diff:
E           {
E            '#DW jobdw capacity=100GB',
E            '#DW stage_in source=/foo',
E         -  '#OAR --account=spam',
E            '#OAR --gres=gpu:4',
E            '#OAR -E '
E            '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.err',
E            '#OAR -O '
E            '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.out',
E         -  '#OAR -l /host=8/core=2,walltime=0:5:0',
E         +  '#OAR -l /host=8/core=2,walltime=0:5:0#OAR --account=spam',
E         ?                                        +++++++++++++++++++
E            '#OAR -n "testjob"',
E           }

```

**Review by vkarak — APPROVED** [review]
Lgtm now. I fixed the unit tests.

File: docs/tutorial_tips_tricks.rst (hunk 5 above)
Comment (mahendrapaipuri) [reply docs/tutorial_tips_tricks.rst:558]: Ok. Will do that in the future PRs.

**Comment by mahendrapaipuri** [comment]
Thanks @vkarak for fixing unit tests.

**Pull request merged**